kratos/utilities/dof_utilities/dof_array_utilities.h

```diff
@@ -0,0 +1,151 @@
+
+    /// Auxiliary slave to master(s) map for slave DOFs
+    using SlaveToMasterDofsMap = std::unordered_map<typename Node::DofType::Pointer, DofsVectorType>;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    DofArrayUtilities() = delete;
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief Set the Up Dof Array object
+     * This function fills a sorted DOFs array with the DOFs of the provided model part (elements, conditions and constraints)
+     * @param rModelPart The model part containing the entities from which the DOFs are to be taken
+     * @param rDofArray The DOFs array to be filled
+     * @param EchoLevel The echo level (i.e., verbosity level)
+     */
+    static void SetUpDofArray(
+        const ModelPart& rModelPart,
+        DofsArrayType& rDofArray,
+        const unsigned int EchoLevel = 0);
+
+    /**
+     * @brief Set the Up Effective Dof Array and Slave To Master Dofs objects
+     * This function fills a map relating each slave DOF with a vector containing its master DOF(s)
+     * If there are constraints, the effective (i.e., non-slave) DOF array is set based on this map
+     * If there are no constraints, the effective DOF array is the standard DOF array
+     * @param rModelPart The model part containing the constraints
+     * @param rDofArray The already filled and sorted DOF array
+     * @param rEffectiveDofArray The effective DOF array to be filled
+     * @param rSlaveToMasterDofsMap The slave to master DOFs map to be filled
+     * @param EchoLevel The echo level (i.e., verbosity level)
+     */
+    static void SetUpEffectiveDofArray(
+        const ModelPart& rModelPart,
+        const DofsArrayType& rDofArray,
+        DofsArrayType& rEffectiveDofArray,
+        SlaveToMasterDofsMap& rSlaveToMasterDofsMap,
+        const unsigned int EchoLevel = 0);
+
+    /**
+     * @brief Set the Dof Equation Ids
+     * This function sets the equation id for each DOF in the given DOFs array
+     * @param rDofArray The already filled and sorted DOF array
+     */
+    static void SetDofEquationIds(const DofsArrayType& rDofArray);
+
+    /**
+     * @brief Set the Effective Dof Equation Ids
+     * This function sets the effective equation ids in the effective DOF array
+     * If the effective DOF array matches the DOF array the effective equation ids are set as the equation ids
+     * @param rDofArray The already filled and sorted DOF array
+     * @param rEffectiveDofArray The effective DOFs array in which the effective DOF id are set
+     */
+    static void SetEffectiveDofEquationIds(
+        const DofsArrayType& rDofArray,
+        DofsArrayType& rEffectiveDofArray);
+
+    ///@}
+    ///@name Input and output
+    ///@{
+
+
+    ///@}
+private:
+    ///@name Member Variables
+    ///@{
+
+
+    ///@}
+    ///@name Private Operations
+    ///@{
+
+
+    ///@}
+}; // Class DofArrayUtilities
+
+///@}
+///@name Input and output
+///@{
+
+
+///@}
```

Comment:
I would not put empty blocks which does not add value, it slows down doxygen processing time.

```suggestion
```